src/pyhf/contrib/viz/brazil.py

```diff
@@ -18,33 +279,115 @@ def plot_results(ax, mutests, tests, test_size=0.05):
-        mutests (:obj:`list` or :obj:`array`): The values of the POI where the
+        test_pois (:obj:`list` or :obj:`array`): The values of the POI where the
           hypothesis tests were performed.
-        tests (:obj:`list` or :obj:`array`): The :math:`\\mathrm{CL}_{s}` values
-          from the hypothesis tests.
-        test_size (:obj:`float`): The size, :math:`\\alpha`, of the test.
+        tests (:obj:`list` or :obj:`array`): The collection of :math:`p`-value-like
+          values (:math:`\mathrm{CL}_{s}` values or tail probabilities) from
+          the hypothesis tests.
+          If the ``components`` keyword argument is ``True``,  ``tests`` is required
+          to have the same structure as :func:`pyhf.infer.hypotest`'s return
+          when using ``return_expected_set=True`` and ``return_tail_probs=True``:
+          a tuple of :math:`\mathrm{CL}_{s}`,
+          :math:`\left[\mathrm{CL}_{s+b}, \mathrm{CL}_{b}\right]`,
+          :math:`\mathrm{CL}_{s,\mathrm{exp}}` band.
+        test_size (:obj:`float`): The size, :math:`\alpha`, of the test.
+        ax (:obj:`matplotlib.axes.Axes`): The matplotlib axis object to plot on.
+
+    Returns:
+        :class:`BrazilBandArtist`: Artist containing the :obj:`matplotlib.artist`
+        objects drawn.
     """
+    if ax is None:
+        ax = plt.gca()
+
+    plot_components = kwargs.pop("components", False)
+    no_cls = kwargs.pop("no_cls", False)
+
+    if plot_components and len(tests[0]) != 3:
+        raise ValueError(
+            f"The components of 'tests' should have len of 3 to visualize the CLs components but have len {len(tests[0])}."
+            + "\n'tests' should have format of: [CLs_obs, [CLsb, CLb], [CLs_exp band]]"
+        )
+
     cls_obs = np.array([test[0] for test in tests]).flatten()
-    cls_exp = [np.array([test[1][i] for test in tests]).flatten() for i in range(5)]
-    ax.plot(mutests, cls_obs, c='black')
-    for idx, color in zip(range(5), 5 * ['black']):
-        ax.plot(
-            mutests, cls_exp[idx], c=color, linestyle='dotted' if idx != 2 else 'dashed'
+    if len(tests[0]) == 3:
+        # split into components
+        tail_probs = np.array([test[1] for test in tests])
+        CLs_exp_set = np.array([test[2] for test in tests])
+    else:
+        CLs_exp_set = np.array([test[1] for test in tests])
+    cls_exp = [
+        np.array([exp_set[sigma_idx] for exp_set in CLs_exp_set]).flatten()
+        for sigma_idx in range(5)
+    ]
+
+    if not no_cls:
+        brazil_band_collection = plot_brazil_band(
+            test_pois, cls_obs, cls_exp, test_size, ax, **kwargs
         )
-    ax.fill_between(mutests, cls_exp[0], cls_exp[-1], facecolor='yellow')
-    ax.fill_between(mutests, cls_exp[1], cls_exp[-2], facecolor='green')
-    ax.plot(mutests, [test_size] * len(mutests), c='red')
-    ax.set_ylim(0, 1)
+    else:
+        # TODO: Find more elegant solution
+        brazil_band_collection = BrazilBandCollection()
+
```

Comment:
perhaps the more elegant solutiion would be to make `plot_cls_components` and `plot_brazil_band` behave similarly and have `plot_brazil_band` also return a list of artists

```
brazil_band_artists = plot_brazil_band(...)
component_artists = plot_cls_components(test_pois, tail_probs, ax, **kwargs)
BrazilBandCollection(*(brazil_band_artists + component_artists))
```

Reply:
The control flow is still there with the changes I added (can't really get away from that), but I agree that using generators makes it look more Pythonic and readable. :+1: